src/box/vinyl.c

```diff
@@ -3225,6 +3264,9 @@ vy_gc(struct vy_env *env, struct vy_recovery *recovery,
 			if (loops % VY_YIELD_LOOPS == 0)
 				fiber_sleep(0);
 		}
+
+		if (lsm_needs_gc)
```

Comment:
`vy_run_remove_files()` is also called by vinyl scheduler on removing compacted run files. I don't think that it is a correct place to put the code that unlinks the space root directory (it is the case only for a fully deleted space).

Reply:
Why not? If there's no run files left after compaction, it looks natural to drop the directory as well, even if the space still exists. The directory will be created on demand, when a new run file is written (see `xlog_create` -> `mkdirpath`).

Reply:
1. When we create a new vinyl space we produce root directory. The idea to remove any evidence of the vinyl space from the file system and then recover the directories while writing to WAL seems non-intuitive to me.
2. The name and the signature of the `vy_run_remove_files(const char *dir, uint32_t space_id, uint32_t iid, int64_t run_id)` function means that in must remove the files related to a specific run, not the root directory where all the runs live.
3. It is additional work to unlink/link root directory on compaction.

I still don't understand why this solution looks simpler then the current one.

Reply:
1. The code that creates an index directory isn't required since commit c00ba8e76470e4dfe131220aaf03cb24c660f438. I think we can drop it now.
2. If a directory is empty, there's no point in keeping it so it can be removed by `vy_run_remove_files`.
3. This work is negligible in comparison to compaction.

> I still don't understand why this solution looks simpler then the current one.

Because it's difficult to comprehend how your solution actually works. Consider this example:

```Lua
box.cfg{checkpoint_count = 1}
box.schema.create_space('test', {engine = 'vinyl', id = 512})
box.space.test:create_index('pk')
box.space.test:insert{1}
box.snapshot()
box.space.test:drop()
box.schema.create_space('test', {engine = 'vinyl', id = 512})
box.space.test:create_index('pk')
box.snapshot()
```

Directory "512/0" will be dropped as a result of this script but space 512 exists.

I don't see any point in trying to describe or fix such corner cases while we can simply say that the space/index directory is removed as soon as there's no files left in it.

Reply:
Ok, as far as you insist, here is  a new version. Now we always remove the root directory on GC (space deletion and compaction). I've preferred to keep functions simple and didn't change `vy_run_remove_files()`. Instead, a new `vy_root_remove_files()` function with similar signature was introduces. It is used to remove the empty directorie files now as far as I don't want to make `vy_run_remove_files()` overcomplicated.

Reply:
Why would the directory removal code suddently make `vy_run_remove_files()` overcomplicated?

If directories were removed in `vy_run_remove_files()`, all the code dealing with file removal would be gathered in one place, not scattered around garbage collection and compaction scheduling.

Reply:
Ok, this is the solution you want to see (if I understand you right). But I am not a big fan of trying to remove the directory for each run. All these attempts would fail (except the last one) and it seems to me a non-optimal code.